dot/doom.d/config.org
#+TITLE: Doom Emacs config
#+AUTHOR: Shahin Sorkh
#+EMAIL: sorkh.shahin@hotmail.com

* Table of contents :TOC:
  - [[#email-accounts][Email accounts]]
- [[#preferences][Preferences]]
  - [[#fonts][Fonts]]
  - [[#misc][Misc]]
- [[#tools][Tools]]
  - [[#local-variables][Local variables]]
  - [[#org-mode][Org-mode]]
  - [[#projectile][Projectile]]
  - [[#company][Company]]
  - [[#postman-import][Postman import]]
  - [[#formatter][Formatter]]
  - [[#super-save][Super save]]
- [[#programming-languages][Programming languages]]
  - [[#php][PHP]]
  - [[#graphql][GraphQL]]
- [[#emacs-application-framework-eaf][Emacs Application Framework /(EAF)/]]
- [[#work][Work]]
  - [[#malltina][Malltina]]

** Email accounts
*** INIT
#+begin_src conf-space :tangle ~/.mbsyncrc
# These settings can also go under an "Account" section, but then the
# "IMAPStore" section needs to reference the account. This is not
# needed for just one store per account.
Create Both
Expunge Both
SyncState *
Sync All
#+end_src
*** Gmail
#+begin_src conf-space :tangle ~/.mbsyncrc
IMAPAccount gmail
Host imap.gmail.com
User ali.zakeri2008@gmail.com
PassCmd "gpg --quiet --decrypt ~/.authinfo.gpg | awk '/machine imap.gmail.com/ {print $NF}'"
SSLType IMAPS
SSLVersion TLSv1.3 TLSv1.2
AuthMechs LOGIN

IMAPStore gmail-remote
Account gmail

MaildirStore gmail-local
Path ~/.mail/ali.zakeri2008@gmail.com/
Inbox ~/.mail/ali.zakeri2008@gmail.com/INBOX
SubFolders Verbatim

Channel gmail-default
Far :gmail-remote:
Near :gmail-local:
Patterns INBOX

Channel gmail-gmail
Far :gmail-remote:"[Gmail]/"
Near :gmail-local:
Patterns * !INBOX !Trash !"All Mail" !Spam

Group gmail
Channel gmail-default
Channel gmail-gmail
#+end_src
#+begin_src emacs-lisp
(set-email-account! "ali.zakeri2008@gmail.com"
                    '((mu4e-sent-folder              . "/ali.zakeri2008@gmail.com/Sent Mail")
                      (mu4e-drafts-folder            . "/ali.zakeri2008@gmail.com/Drafts")
                      (mu4e-trash-folder             . "/ali.zakeri2008@gmail.com/Trash")
                      (mu4e-refile-folder            . "/ali.zakeri2008@gmail.com/All Mails")
                      (mu4e-compose-signature        . "\nBest regards,\nShSo")
                      (smtpmail-smtp-server          . "smtp.gmail.com")
                      (smtpmail-smtp-user            . "ali.zakeri2008@gmail.com")
                      (mu4e-compose-signature-auto-include nil))
                    nil)
#+end_src
*** Hotmail
#+begin_src conf-space :tangle ~/.mbsyncrc
IMAPAccount hotmail
Host outlook.office365.com
User sorkh.shahin@hotmail.com
PassCmd "gpg --quiet --decrypt ~/.authinfo.gpg | awk '/machine outlook.office365.com/ {print $NF}'"
SSLType IMAPS
SSLVersion TLSv1.3 TLSv1.2
AuthMechs LOGIN

IMAPStore outlook-remote
Account hotmail

MaildirStore outlook-local
Path ~/.mail/sorkh.shahin@hotmail.com/
Inbox ~/.mail/sorkh.shahin@hotmail.com/INBOX
SubFolders Verbatim

Channel outlook
Far :outlook-remote:
Near :outlook-local:
Patterns % !Archives !Deleted !Junk
#+end_src
#+begin_src emacs-lisp
(set-email-account! "sorkh.shahin@hotmail.com"
                    '((mu4e-sent-folder              . "/sorkh.shahin@hotmail.com/Sent")
                      (mu4e-drafts-folder            . "/sorkh.shahin@hotmail.com/Drafts")
                      (mu4e-trash-folder             . "/sorkh.shahin@hotmail.com/Deleted")
                      (mu4e-refile-folder            . "/sorkh.shahin@hotmail.com/All")
                      (mu4e-compose-signature        . "\nBest regards,\nShSo")
                      (smtpmail-smtp-server          . "smtp.office365.com")
                      (smtpmail-smtp-user            . "sorkh.shahin@hotmail.com")
                      (mu4e-compose-signature-auto-include t))
                    t)
#+end_src
* Preferences
** Fonts
#+begin_src emacs-lisp
(setq doom-font (font-spec :family "JetBrains Mono" :size 13)
      doom-unicode-font (font-spec :family "Vazir Code" :size 11)
      doom-big-font (font-spec :family "JetBrains Mono" :size 16)
      doom-variable-pitch-font (font-spec :family "Overpass" :size 14))
#+end_src
** Misc
#+begin_src emacs-lisp
(setq doom-theme 'doom-one-light)
(setq display-line-numbers-type 'visual)
(setq treemacs-git-mode 'deffered)
#+end_src
* Tools
** Local variables
#+begin_src emacs-lisp
(setq enable-local-variables t)
#+end_src
** Org-mode
*** General
#+begin_src emacs-lisp
(setq org-directory "~/org/"
      org-passwords-file "~/.authinfo.gpg"
      org-use-property-inheritance t)
(setq deft-directory "~/org/"
      deft-recursive t)
;; Replace the content marker, “⋯”, with a nice unicode arrow.
(setq org-ellipsis " ⤵")
;; Fold all source blocks on startup.
(setq org-hide-block-startup t)
;; Lists may be labelled with letters.
(setq org-list-allow-alphabetical t)
;; Avoid accidentally editing folded regions, say by adding text after an Org “⋯”.
(setq org-catch-invisible-edits 'show)
;; Tangling should preserve my indentation.
(setq org-src-preserve-indentation t)
;; Tab should do indent in code blocks
(setq org-src-tab-acts-natively t)
;; Give quote and verse blocks a nice look.
(setq org-fontify-quote-and-verse-blocks t)
;; Pressing ENTER on a link should follow it.
(setq org-return-follows-link t)
#+end_src
*** Journal
#+begin_src emacs-lisp
(after! org
  (setq org-journal-date-format "%y-%m-%d"
        org-journal-file-type "org"
        org-journal-file-format "YYYYMMDD.org"))
#+end_src
*** Notes
#+begin_src emacs-lisp
(after! org
  (setq org-noter-notes-search-path '("~/org/notes/")
        org-default-notes-file "~/org/notes.org"))
(after! org
  (setq +org-capture-frame-parameters
        `((name . "doom-capture")
          (width . 70)
          (height . 25)
          (transient . t)
          ,(when IS-LINUX
             `(display . ,(or (getenv "DISPLAY") ":0")))
          ,(if IS-MAC '(menu-bar-lines . 1)))))
#+end_src
*** Bindings
#+begin_src emacs-lisp
(after! org
  (map! :map org-mode-map
        :n "M-j" #'org-metadown
        :n "M-k" #'org-metaup))
#+end_src
** Projectile
My projects directories is typically as follows:
#+begin_example
~/code
├── github.com
│   └── {username}
│       └── {repo-name}
├── gitlab.com
│   └── {repo-namespace}
│       └── {repo-name}
├── {private-repo}
│   └── {repo-namespace}
│       └── {repo-name}
└── localhost
    └── {project-name}
#+end_example
So I need a method to help =projectile= discover them all.
~my/find-directories-recursive~ is my attempt to achieve this.
#+begin_src emacs-lisp :tangle autoload.el
(defun my/is-dir (file-attrs)
  (and
   (nth 1 file-attrs)
   (not (cl-search "/node_modules" (my/get-file-path file-attrs)))
   (not (cl-search "/vendor" (my/get-file-path file-attrs)))
   (not (cl-search "/.git" (my/get-file-path file-attrs)))))

(defun my/get-file-path (file-attrs)
  (nth 0 file-attrs))

(defun my/list-directories (directory)
  (cl-map 'list 'my/get-file-path
          (seq-filter 'my/is-dir
                      (cddr (directory-files-and-attributes directory t)))))

;;;###autoload
(defun my/find-directories-recursive (directory depth)
  "Recursively goes into `DIRECTORY' and returns all dirs found."
  (let ((found-dirs (my/list-directories directory)))
    (if (> depth 0)
        (flatten-list (merge 'list found-dirs (cl-map 'list
                                                      #'(lambda (d) (my/find-directories-recursive d (- depth 1)))
                                                      found-dirs)
                             '(lambda (&rest ...) t)))
      found-dirs)))
#+end_src
#+begin_src emacs-lisp
(after! projectile
  (setq projectile-project-search-path (my/find-directories-recursive "~/code" 3))
  (setq projectile-auto-discover t))
#+end_src
** Company
Still struggling with ~TAB~ being trigger for =company= /and/ =yasnippet= expansion.
#+begin_src emacs-lisp
(after! company
  (after! company-tabnine
    (setq +lsp-company-backends '(:separate company-tabnine company-capf)))
  (setq company-idle-delay 0.4)
  (setq company-tooltip-width-grow-only t)
  (setq company-show-quick-access t))
#+end_src

*** TabNine
Because I like TabNine.
#+begin_src emacs-lisp :tangle packages.el
(when (featurep! :completion company)
  (package! company-tabnine))
#+end_src
** Postman import
#+begin_src emacs-lisp :tangle packages.el
(package! impostman
  :recipe (:host github :repo "flashcode/impostman"))
#+end_src
** Formatter
#+begin_src emacs-lisp
(setq +format-on-save-enabled-modes
      '(not emacs-lisp-mode  ; elisp's mechanisms are good enough
            haskell-mode     ; it's not configured (yet)
            sql-mode         ; sqlformat is currently broken
            tex-mode         ; latexindent is broken
            latex-mode))
#+end_src
** Super save
#+begin_src emacs-lisp :tangle packages.el
(package! super-save)
#+end_src
#+begin_src emacs-lisp
(after! super-save
  (super-save-mode +1)
  (setq super-save-auto-save-when-idle t)
  (setq auto-save-default nil)
  (setq super-save-remote-files nil)
  (setq super-save-exclude '(".gpg")))
#+end_src
* Programming languages
** PHP
*** template file
Fix php file template error.
#+begin_src emacs-lisp :tangle autoload.el
;;;###autoload
(defun doom/php-class-name ()
  (replace-regexp-in-string "\\.php$" "" (buffer-name)))
#+end_src
*** tests related
Add =behat= tests support.
#+begin_src emacs-lisp :tangle packages.el
(package! feature-mode)
#+end_src
*** formatter
Use local =php-cs-fixer= config always.
+ [ ] fallback on some default values
+ [ ] search for all possible config file names
#+begin_src emacs-lisp
(after! php
  (setq-hook! 'php-mode-hook php-cs-fixer-config-option
              (concat (doom-project-root) ".php-cs-fixer.dist.php"))
  (setq-hook! 'php-mode-hook php-cs-fixer-rules-level-part-options
              '("@PSR1" "@PSR2" "@PSR12" "@PSR12:risky" "@Symfony" "@Symfony:risky"))
  (setq-hook! 'php-mode-hook php-cs-fixer-rules-fixer-part-options
              '("array_indentation" "explicit_indirect_variable" "method_chaining_indentation" "-yoda_style" "-psr0")))
(after! format-all
  (setq-hook! 'php-mode-hook +format-with-lsp nil)
  (define-format-all-formatter php-cs-fixer
    (:executable)
    (:install "composer require --dev friendsofphp/php-cs-fixer")
    (:modes php-mode)
    (:format
     (format-all--buffer-native 'php-mode 'php-cs-fixer-fix))))
#+end_src
** GraphQL
#+begin_src emacs-lisp :tangle packages.el
(package! graphql-mode)
#+end_src
* Emacs Application Framework /(EAF)/
I don't know how to make this guy run (yet)!
#+begin_src emacs-lisp :tangle packages.el
;; (package! eaf
;;     :recipe (:local-repo "modules/eaf"
;;                 :files ("*.el" "core/*.el" "extension/*.el" "app/*/*.el" "*.py")))

;; (when (package! eaf :recipe (:host github
;;                              :repo "manateelazycat/emacs-application-framework"
;;                              :files ("*.el" "*.py" "app" "core")
;;                              :build (:not compile)))
;;     (package! ctable :recipe (:host github :repo "kiwanami/emacs-ctable"))
;;     (package! deferred :recipe (:host github :repo "kiwanami/emacs-deferred"))
;;     (package! epc :recipe (:host github :repo "kiwanami/emacs-epc")))

;; (use-package! eaf
;;   :commands (eaf-open-browser eaf-open find-file)
;;   :config
;;   (use-package! ctable)
;;   (use-package! deferred)
;;   (use-package! epc))
#+end_src
#+begin_src emacs-lisp
;; (after! eaf
;;     (require 'eaf-browser)
;;     (require 'eaf-image-viewer)
;;     (require 'eaf-markdown-previewer)
;;     (require 'eaf-mindmap)
;;     (require 'eaf-music-player)
;;     (require 'eaf-org-previewer)
;;     (require 'eaf-pdf-viewer)
;;     (require 'eaf-video-player))
#+end_src
* Work
** Malltina
#+begin_src conf-space :tangle ~/.mbsyncrc
IMAPAccount malltina
Host mail.malltina.com
User shahin@malltina.com
PassCmd "gpg --quiet --decrypt ~/.authinfo.gpg | awk '/machine mail.malltina.com/ {print $NF}'"
SSLType STARTTLS
CertificateFile ~/.mail/shahin@malltina.com/ca.cert

IMAPStore malltina-remote
Account malltina

MaildirStore malltina-local
Path ~/.mail/shahin@malltina.com/
Inbox ~/.mail/shahin@malltina.com/INBOX
SubFolders Verbatim

Channel malltina
Far :malltina-remote:
Near :malltina-local:
Patterns % !"Deleted Items"
#+end_src
#+begin_src emacs-lisp
(after! forge
  (add-to-list 'forge-alist '("git.malltina.com"
                              "git.malltina.com/api/v4"
                              "git.malltina.com"
                              forge-gitlab-repository)))
(set-email-account! "shahin@malltina.com"
                    '((mu4e-sent-folder              . "/shahin@malltina.com/Sent Items")
                      (mu4e-drafts-folder            . "/shahin@malltina.com/Drafts")
                      (mu4e-trash-folder             . "/shahin@malltina.com/Deleted Items")
                      (mu4e-refile-folder            . "/shahin@malltina.com/All Items")
                      (mu4e-compose-signature        . "\nBest regards,\nShSo")
                      (smtpmail-smtp-server          . "mail.malltina.com")
                      (smtpmail-smtp-user            . "shahin@malltina.com")
                      (mu4e-compose-signature-auto-include t))
                    nil)
#+end_src


dot/doom.d/config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!


;; Some functionality uses this to identify you, e.g. GPG configuration, email
;; clients, file templates and snippets.
(setq user-full-name "Shahin Sorkh"
      user-mail-address "sorkh.shahin@hotmail.com")

(set-email-account! "sorkh.shahin@hotmail.com"
                    '((mu4e-sent-folder                 . "/sorkh.shahin@hotmail.com/Sent")
                      (mu4e-drafts-folder            . "/sorkh.shahin@hotmail.com/Drafts")
                      (mu4e-trash-folder             . "/sorkh.shahin@hotmail.com/Deleted")
                      (mu4e-refile-folder            . "/sorkh.shahin@hotmail.com/All")
                      (mu4e-compose-signature        . "\nBest regards,\nShSo")
                      (smtpmail-smtp-server          . "smtp.office365.com")
                      (smtpmail-smtp-user            . "sorkh.shahin@hotmail.com")
                      (mu4e-compose-signature-auto-include t))
                    t)
(set-email-account! "ali.zakeri2008@gmail.com"
                    '((mu4e-sent-folder                 . "/ali.zakeri2008@gmail.com/Sent Mail")
                      (mu4e-drafts-folder            . "/ali.zakeri2008@gmail.com/Drafts")
                      (mu4e-trash-folder             . "/ali.zakeri2008@gmail.com/Trash")
                      (mu4e-refile-folder            . "/ali.zakeri2008@gmail.com/All Mails")
                      (mu4e-compose-signature        . "\nBest regards,\nShSo")
                      (smtpmail-smtp-server          . "smtp.gmail.com")
                      (smtpmail-smtp-user            . "ali.zakeri2008@gmail.com")
                      (mu4e-compose-signature-auto-include nil))
                    nil)
(set-email-account! "shahin@malltina.com"
                    '((mu4e-sent-folder                 . "/shahin@malltina.com/Sent Items")
                      (mu4e-drafts-folder            . "/shahin@malltina.com/Drafts")
                      (mu4e-trash-folder             . "/shahin@malltina.com/Deleted Items")
                      (mu4e-refile-folder            . "/shahin@malltina.com/All Items")
                      (mu4e-compose-signature        . "\nBest regards,\nShSo")
                      (smtpmail-smtp-server          . "mail.malltina.com")
                      (smtpmail-smtp-user            . "shahin@malltina.com")
                      (mu4e-compose-signature-auto-include t))
                    nil)

;; Doom exposes five (optional) variables for controlling fonts in Doom. Here
;; are the three important ones:
;;
;; + `doom-font'
;; + `doom-variable-pitch-font'
;; + `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;;
;; They all accept either a font-spec, font string ("Input Mono-12"), or xlfd
;; font string. You generally only need these two:
(setq doom-font (font-spec :family "JetBrains Mono" :size 13)
      doom-unicode (font-spec :family "Vazir Mono" :size 13)
      doom-big-font (font-spec :family "JetBrains Mono" :size 16)
      doom-variable-pitch-font (font-spec :family "Overpass" :size 14))

;; There are two ways to load a theme. Both assume the theme is installed and
;; available. You can either set `doom-theme' or manually load a theme with the
;; `load-theme' function. This is the default:
(setq doom-theme 'doom-horizon)

;; If you use `org' and don't want your org files in the default location below,
;; change `org-directory'. It must be set before org loads!
(setq org-directory "~/org/")
(setq org-noter-notes-search-path '("~/org/notes/"))
(setq deft-directory "~/org/"
      deft-recursive t)
(setq org-journal-date-format "%y-%m-%d"
      org-journal-file-type "org"
      org-journal-file-format "YYYYMMDD.org")

;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
(setq display-line-numbers-type 'relative)


;; Here are some additional functions/macros that could help you configure Doom:
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package!' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys
;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
;; This will open documentation for it, including demos of how they are used.
;;
;; You can also try 'gd' (or 'C-c c d') to jump to their definition and see how
;; they are implemented.

(when (featurep! :editor format)
  (add-hook 'before-save-hook #'+format/buffer))

(after! projectile
  (setq projectile-project-search-path (my/find-directories-recursive "~/code" 3))
  (setq projectile-auto-discover t))

(after! company
  (setq +lsp-company-backends '(:separate company-tabnine company-capf :with company-yasnippet))
  (setq company-idle-delay 0.4)
  (setq company-tooltip-width-grow-only t)
  (setq company-show-quick-access t))

(after! org
  (map! :map org-mode-map
        :n "M-j" #'org-metadown
        :n "M-k" #'org-metaup))

(after! php
  (setq-hook! php-mode php-cs-fixer-config-option
              (concat (projectile-project-root) ".php-cs-fixer.dist.php")))

(after! format-all
  (define-format-all-formatter php-cs-fixer
    (:executable)
    (:install "composer require --dev friendsofphp/php-cs-fixer")
    (:modes php-mode)
    (:format
     (format-all--buffer-native 'php-mode #'php-cs-fixer-fix))))

;; (setf (alist-get 'markdown-mode +spell-excluded-faces-alist)
;;       '(markdown-code-face
;;         markdown-reference-face
;;         markdown-link-face
;;         markdown-url-face
;;         markdown-markup-face
;;         markdown-html-attr-value-face
;;         markdown-html-attr-name-face
;;         markdown-html-tag-name-face))

